### DISCUSSION TITLE
Machine Learning Nodes

### DISCUSSION THREAD

**Discussion by jackcarey** [GimelStudio/GimelStudio#156] [Ideas]
Hi, I'm new to the project and it's exactly what I'm looking for so I'm keen to start developing some custom machine learning based nodes *(timeframe TBC...)*. I'm thinking about:

|Model|Input|Output|Purpose
|:--|:-:|:-:|:--
|[Stable Diffusion text-to-image](https://huggingface.co/spaces/stabilityai/stable-diffusion)|1x text|1x image|Generate an image from text input
|[Depth Estimation](https://huggingface.co/spaces/Detomo/Depth-Estimation)|1x image|1x image|output a gradient image of the estimated depth of pixels
|[Object Segmentation](https://huggingface.co/tasks/image-segmentation)|1x image| Nx images| automatically divide an image into separate layers/masks based on its content, so each can be handled separately.
|[Style Transfers](https://huggingface.co/spaces/aravinds1811/neural-style-transfer)|2x images|1x image| easily change the style of images

Possible ancillary nodes needed:
- 'Text' node, like 'Value' but for strings instead of ints
- Allow 'Value' to output floats, for tweaking ML parameters

Questions: 
 - When I start, am I okay to work from the re-write in [PR155](https://github.com/GimelStudio/GimelStudio/pull/155), or should I stick to cloning the repo?
 - What do you think of the ideas, are there any other models that could be useful?

**Comment by iwoithe** [comment]
Thanks for your interest in contributing. :slightly_smiling_face:

> timeframe TBC...

No problem. No time pressure at all. :slightly_smiling_face:

> 'Text' node, like 'Value' but for strings instead of ints

Do you mean https://github.com/GimelStudio/GimelStudio/blob/master/src/nodes/corenodes/input/string_node.py?

> Allow 'Value' to output floats, for tweaking ML parameters

I must admit I forgot we only had an integer node (https://github.com/GimelStudio/GimelStudio/blob/master/src/nodes/corenodes/input/integer_node.py). <strike>Working on a float node now</strike>. See https://github.com/GimelStudio/GimelStudio/pull/157 (now merged).

> When I start, am I okay to work from the re-write in https://github.com/GimelStudio/GimelStudio/pull/155, or should I stick to cloning the repo?

Use the `master` branch.

There will be an announcement coming in the near-ish future. There is barely anything in the `develop` branch currently, as shown by the relatively up-to-date screenshots in that PR.

> What do you think of the ideas, are there any other models that could be useful?

I must admit I haven't done too much regarding machine learning, but it sounds like it would be nice. What you've listed seems to be relatively useful.

> start developing some custom machine learning based nodes

Go for it. :slightly_smiling_face:

**Comment by yonMaor** [comment]
Are we still working on this?
I'd love to get involved

**Comment by jackcarey** [comment]
Life has gotten in the way a bit, but one approach I have considered is a
node that can detect and interact with something like Automatic1111 for
Stable Diffusion. This wouldn’t be the end goal of the nodes as it isn’t
low-code enough, but it’s a starting point that power users could take
advantage of

**Comment by Correct-Syntax** [comment]
My thought on this would be that at least the image generation would be out of scope for the Gimel Studio core nodes. 

The sd image generation and perhaps even the other ones relying on models could be a separate node that users could install, but maybe not something to include in the "core" nodes shipped with Gimel Studio.

I'm not really aware of how this works on a practical level though. Is it an api that does this or do you have to have a copy of the ml model on your computer in order for it to work?

**Reply by jackcarey** [reply]
Automatic1111/SD spins up a server, so the idea with that is that people on the same network could queue up jobs on a shared instance with decent hardware, but hobbyists just run it locally. It also uses OpenAPI to describe the endpoints available, so this could be used to automatically validate user configuration.

Yes, a copy of the model would have to be available wherever Automatic1111 or the node executes. This could become part of the execution pipeline, but long download times are detrimental to UX and potentially confusing.

Agreed it isn't a core node, but one to be added by users who know what they're doing.

**Comment by yonMaor** [comment]
I imagine that we could give users the ability to add their own nodes (similar to blender addons). 
This could be a good opportunity to lay the foundations for a feature like this

**Comment by iwoithe** [comment]
> I imagine that we could give users the ability to add their own nodes (similar to blender addons).

Correct me if I'm wrong @Correct-Syntax, but support for custom nodes already exist. However, add-ons (as in, adding additional functionality to the app *not including nodes*) is not implemented.

**Comment by jackcarey** [comment]
If A1111 is the chosen route for ML nodes then this could actually go in as an 'OpenAPI'/ 'HTTP'/ 'Fetch' node instead. This change would decouple from having to download models at all and could bring in other remote services - it'd make sense as a core node?

For ML itself, I still think having nodes capable of handling the downloads& execution is useful, and the recent Hugging Face Transformer Agents provide a useful API for that, on top of the existing Transformers library. See: https://huggingface.co/docs/transformers/transformers_agents

These come with the caveat that image generation updates might then be better behind a user-triggered button, instead of on-change (due to time/memory/network use). Not ideal, but if nodes could have a 'forceFlowToManualGeneration' bool on them, the best of both could be achieved?

**Comment by jackcarey** [comment]
I'm unsure where GS is up to with multithreading but the Python 3.12 sub-interpreter features are exciting for true asynchronous tasks. It's still a few months out but this may well become necessary to use for Fetch and ML nodes.

Release notes: https://docs.python.org/3.12/whatsnew/3.12.html
